### ISSUE TITLE
questions about 0.11 resource semantics and design

### ISSUE THREAD

**Issue by softprops** [hyperium/hyper#1133]
I've been using hyper for a while now and am trying to equip myself to understand some of the changes coming down the pipe with the future 0.11.x versions. I had a few questions.

1. In the past, I would store a reference to a hyper::Client in a struct of sorts and decorate that type with methods adding behavior. I did this because as I understood, Clients acquire resources and it was wasteful to create a new Client for each request execution. Is that no longer the case? It looks like you pass in the resources a client uses to do its execution in the newer Client constructor, typically a handle from a tokio core.  Does it still make sense to pass around a struct with a reference to a Client or is it a better design to instead just store a handle to a tokio Core and create a new Client on each request.

Also I found it bit awkward to update my original designs with a new type parameter

before

```rust
struct MyLib {
  client: Client
}

impl MyLib {

}
```

after
```rust
struct MyLib<C> {
  client: Client<C>
}

impl<C: Connect> MyLib<C> {
}
```

2. In the new [server example](https://github.com/hyperium/hyper/blob/f05a58a1b2288d30a601b0466bc08a38c5a76054/examples/server.rs#L52), the logging mentions that only 1 thread is in use. Are there any caveats to this or would a typical server need more? Im asking because I saw [this article](https://blog.guillaume-gomez.fr/articles/2017-02-22+Rust+asynchronous+HTTP+server+with+tokio+and+hyper) and wasn't sure the what kind of throughput to expect with the out of the box `Http::new().bind` interface.

**Comment by mgattozzi** [comment]
I don't know if it's wasteful still but I look at it like I'm constructing a future to be executed on an event loop (in this case Tokio Core). In my Github library I pass around a reference to the core construct the Future based off the Request and then send it off. You can see the bit of code that does that [here](https://github.com/mgattozzi/github-rs/blob/master/src/client.rs#L352-L383) No need to pass Client<C> around or anything. You could just pass the Core around in the Struct instead. That being said I don't know if this is the best way to do it but I've not found any issues with it personally.

I'm not as sure about number 2.

**Comment by seanmonstar** [comment]
1. No, your previous behavior is still valid. The `Client` takes a `&Handle`, so that it can attach to a tokio `Core`, but owns its own resources. If you keep constructing new `Client`s and throwing them away, you will be constructing new DNS worker threadpools and connection pools and throwing them away, so try not to do that.

  Regarding the type parameter, that's to allow a `Client` to have different connectors. Previously, the type parameter was removed by keeping a boxed trait object, but that's not what hyper should be doing. By having the generic, someone can now choose to box it or not. If you're building a library, it may make sense to be generic also. If you're using a client in your application, you probably know what connector you will be using.

2. The logging is just to be clear, since some people would start it up and assume it was using all their cores. That's not necessarily the best default. Each server should determine where it is spending its time. If the thread running the `Core`/`hyper::Server` is at 100% and others are idle, you may want to try spawning more cores in more threads.

**Comment by softprops** [comment]
@seanmonstar thanks for responding. Awesome work as always. I like the extra effort  make to
sure this interface is as ergonomic as it can be. I know there has been a ton of iteration on the async interface. It was worth it. Looking forward to the next release.

@mgattozzi thanks for the link. I'm actually in the procress of refactoring my [github client](https://github.com/softprops/hubcaps/blob/master/src/lib.rs#L188) to use the master branch of hyper before it's release. I've got a lot of work to do! But I really like what I'm seeing so far.

Closing this issue since I think my questions were answered.

Thanks guys!
